// @wppconnect/server/dist/util/tokenStore/mongodbTokenStory.js
"use strict";var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.default = void 0;var _token = _interopRequireDefault(require("./model/token"));

class MongodbTokenStore {

  constructor(client) {
    this.client = client;
  }
  tokenStore = {
    getToken: async (sessionName) => {
      let result = await _token.default.findOne({ sessionName });
      if (result === null) return result;
      result = JSON.parse(JSON.stringify(result));
      result.config = JSON.parse(result.config);
      result.config.webhook = result.webhook;
      this.client.config = result.config;
      return result;
    },
    setToken: async (sessionName, tokenData) => {
      const token = new _token.default(tokenData);
      token.sessionName = sessionName;
      token.webhook = this.client.config.webhook;
      token.config = JSON.stringify(this.client.config);

      const tk = await _token.default.findOne({ sessionName });

      if (tk) {
        token._id = tk._id;
        return (await _token.default.updateOne({ _id: tk._id }, token)) ?
        true :
        false;
      } else {
        return (await token.save()) ? true : false;
      }
    },
    removeToken: async (sessionName) => {
      return (await _token.default.deleteOne({ sessionName })) ? true : false;
    },
    listTokens: async () => {
      const result = await _token.default.find();
      return result.map((m) => m.sessionName);
    }
  };
}var _default = exports.default =

MongodbTokenStore;
