// @wppconnect/server/dist/util/tokenStore/redisTokenStory.js
"use strict";var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.default = void 0;var _config = _interopRequireDefault(require("../../config"));
var _db = _interopRequireDefault(require("../db/redis/db"));
var _functions = require("../functions");

class RedisTokenStore {


  constructor(client) {
    this.client = client;

    let prefix = _config.default.db.redisPrefix || '';
    if (prefix === 'docker') {
      prefix = (0, _functions.getIPAddress)();
    }
  }
  tokenStore = {
    getToken: (sessionName) =>
    new Promise((resolve, reject) => {
      _db.default.get(
        this.prefix + sessionName,
        (err, reply) => {
          if (err) {
            return reject(err);
          }
          const object = JSON.parse(reply);
          if (object) {
            if (object.config && Object.keys(this.client.config).length === 0)
            this.client.config = object.config;
            if (
            object.webhook &&
            Object.keys(this.client.config).length === 0)

            this.client.config.webhook = object.webhook;
          }
          resolve(object);
        }
      );
    }),
    setToken: (sessionName, tokenData) =>
    new Promise((resolve) => {
      tokenData.sessionName = sessionName;
      tokenData.config = this.client.config;
      _db.default.set(
        this.prefix + sessionName,
        JSON.stringify(tokenData),
        (err) => {
          return resolve(err ? false : true);
        }
      );
    }),
    removeToken: (sessionName) =>
    new Promise((resolve) => {
      _db.default.del(this.prefix + sessionName, (err) => {
        return resolve(err ? false : true);
      });
    }),
    listTokens: () =>
    new Promise((resolve) => {
      _db.default.keys(this.prefix + '*', (err, keys) => {
        if (err) {
          return resolve([]);
        }
        keys.forEach((item, indice) => {
          if (this.prefix !== '' && item.includes(this.prefix)) {
            keys[indice] = item.substring(
              item.indexOf(this.prefix) + this.prefix.length
            );
          }
        });
        return resolve(keys);
      });
    })
  };
}var _default = exports.default =

RedisTokenStore;
